File: frontend/lib/models/memo.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:frontend/models/tag.dart';

part 'memo.freezed.dart';
part 'memo.g.dart';

extension type const MemoId(int value) {}

@freezed
sealed class Memo with _$Memo {
  @JsonSerializable(fieldRename: FieldRename.snake)
  const factory Memo({
    @MemoIdJsonConverter() required MemoId id,
    required String title,
    required String body,
    required String raw,
    required DateTime createdAt,
    @Default(false) bool isFavorite,
    @Default([]) List<Tag> tags,
    DateTime? updatedAt,
  }) = _Memo;

  factory Memo.fromJson(Map<String, dynamic> json) => _$MemoFromJson(json);

  const Memo._();

  List<String> get tagNames => tags.map((tag) => tag.name).toList();
}

final class MemoIdJsonConverter implements JsonConverter<MemoId, int> {
  const MemoIdJsonConverter();

  @override
  MemoId fromJson(int json) => MemoId(json);

  @override
  int toJson(MemoId object) => object.value;
}

File: backend/tests/mock_data/memo.py
from datetime import datetime

from utils.jst_now import JST

SHORT_MEMOS = [
    {
        "title": "メモ1",
        "user_id": "a",
        "body": "sample",
        "raw": "sample",
        "is_favorite": False,
        "is_archive": False,
        "simple_embedding": [0.1] * 3,
        "created_at": datetime(2024, 4, 1, 15, 30, 0, tzinfo=JST),
        "updated_at": datetime(2024, 4, 9, 15, 30, 0, tzinfo=JST),
    },
    {
        "title": "メモ2",
        "user_id": "a",
        "body": "sample",
        "raw": "sample",
        "is_favorite": False,
        "is_archive": False,
        "simple_embedding": [0.1] * 3,
        "created_at": datetime(2024, 4, 4, 15, 30, 0, tzinfo=JST),
        "updated_at": datetime(2024, 4, 8, 15, 30, 0, tzinfo=JST),
    },
    {
        "title": "Memo 3",
        "user_id": "b",
        "body": "sample",
        "raw": "sample",
        "is_favorite": False,
        "is_archive": False,
        "simple_embedding": [0.1] * 3,
        "created_at": datetime(2024, 4, 3, 15, 30, 0, tzinfo=JST),
        "updated_at": datetime(2024, 4, 3, 15, 30, 0, tzinfo=JST),
    },
    {
        "title": "Memo 4",
        "user_id": "a",
        "body": "sample",
        "raw": "sample",
        "is_favorite": False,
        "is_archive": False,
        "simple_embedding": [0.1] * 3,
        "created_at": datetime(2024, 4, 3, 15, 30, 0, tzinfo=JST),
        "updated_at": datetime(2024, 4, 7, 15, 30, 0, tzinfo=JST),
    },
]

EMPTY_MEMOS = []
